ldms/src/sampler/curated_metrics/curated_metrics.c

```diff
@@ -0,0 +1,1147 @@
+	{ "fork", 1, PROC_STAT, LDMS_V_D64, fork_cb },
+	{ "procs_blocked", 1, PROC_STAT, LDMS_V_D64, procs_block_cb },
+	{ "iface_name", 0, NET_DEV, LDMS_V_CHAR_ARRAY, NULL },
+	{ "rx_bytes", 0, NET_DEV, LDMS_V_D64, rxb_cb },
+	{ "rx_packets", 0, NET_DEV, LDMS_V_D64, rx_packets_cb },
+	{ "rx_errs", 0, NET_DEV, LDMS_V_D64, rx_errs_cb },
+	{ "rx_drop", 0, NET_DEV, LDMS_V_D64, rx_drop_cb },
+	{ "tx_bytes", 0, NET_DEV, LDMS_V_D64, txb_cb },
+	{ "tx_packets", 0, NET_DEV, LDMS_V_D64, txp_cb },
+	{ "tx_errs", 0, NET_DEV, LDMS_V_D64, tx_errs_cb }
+};
+
+enum stat_row {
+	STAT_CPU = 1,
+	STAT_CTXT,
+	STAT_BTIME,
+	STAT_PROCESSES,
+	STAT_PROCS_RUNNING,
+	STAT_PROCS_BLOCKED,
+	STAT_INTR,
+	STAT_SOFTIRQ,
+};
+
+struct stat_row_ent {
+	const char *prefix;
+	enum stat_row type;
+};
+
+struct stat_row_ent stat_row_ents[] = {
+	/* sorted by prefix for bsearch */
+	{ "btime",         STAT_BTIME         },
+	{ "cpu",           STAT_CPU           },
+	{ "ctxt",          STAT_CTXT          },
+	{ "intr",          STAT_INTR          },
+	{ "processes",     STAT_PROCESSES     },
+	{ "procs_blocked", STAT_PROCS_BLOCKED },
+	{ "procs_running", STAT_PROCS_RUNNING },
+	{ "softirq",       STAT_SOFTIRQ       },
+};
+
+int stat_row_cmp(const void *key, const void *_ent)
+{
+	const struct stat_row_ent *ent = _ent;
+	return strcmp(key, ent->prefix);
+}
+
+static int parse_cmetrics(curated_metrics_t mi, char *path) {
+	int i;
+	int rc = 0;
+	json_t *cfp, *cfg_metrics;
+	json_error_t jerr;
+	cfp = json_load_file(path, 0, &jerr);
+	if (!cfp) {
+		ovis_log(mi->log, OVIS_LERROR, SAMP ": Could not open json file %s : exiting sampler\n", path);
+		errno = ENOENT;
+		goto err_1;
+	}
+	cfg_metrics = json_object_get(cfp, "metrics");
+	if (!cfg_metrics) {
+		ovis_log(mi->log, OVIS_LERROR, SAMP ": Key 'metrics' is not present in json file %s\n", path);
+		errno = ENOENT;
+		goto err_1;
+	}
+	for (i = 0; i < ARRAY_LEN(metric_list); i++) {
+		mi->collect_list[i].collect = 0;
+	}
+	if (cfg_metrics && json_is_array(cfg_metrics)) {
+		int num_metrics = 0;
+		int j;
+		num_metrics = json_array_size(cfg_metrics);
+		for (j = 0; j < num_metrics; j++) {
+			const char *mname;
+			mname = json_string_value(json_array_get(cfg_metrics, j));
+			for (i = 0; i < ARRAY_LEN(metric_list); i++) {
+				rc = strcmp(mname, metric_list[i].name);
+				if (!rc) {
+					mi->collect_list[i].collect = 1;
+					break;
+				}
+			}
+		}
+	}
+	return 0;
+err_1:
+	json_decref(cfp);
+	return rc;
+}
+
+/* end loadavg */
+static int create_metric_set(curated_metrics_t mi)
+{
+	ldms_schema_t schema;
+	int rc, i;
+	rc = 0;
+
+	schema = base_schema_new(mi->base);
+	if (!schema) {
+		ovis_log(mi->log, OVIS_LERROR,
+		       "%s: The schema '%s' could not be created, errno=%d.\n",
+		       __FILE__, mi->base->schema_name, errno);
+		rc = errno;
+		goto err;
+	}
+
+	mi->mf = fopen(procmeminfo, "r");
+	if (!mi->mf) {
+		ovis_log(mi->log, OVIS_LERROR, "Could not open the " SAMP " file "
+				"'%s'...exiting sampler\n", procmeminfo);
```

Comment:
A nit .. when `fopen()` returns NULL, it sets `errno`. You could include `strerror(errno)` in the log message to provide more information to users.